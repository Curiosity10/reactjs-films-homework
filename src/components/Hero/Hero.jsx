import React, { useState, memo } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames/bind';
import styles from './Hero.scss';
import StarRating from '../StarRating';
import NumRating from '../NumRating';
import Genres from '../Genres';
import Button from '../Button';
import Spinner from '../Spinner';

const cx = cn.bind(styles);

const Hero = ({
  movie,
  fetchVideoKey,
  toggleModal,
  isLoading,
}) => {
  const [showInfo, setShowInfo] = useState(true);

  const {
    title,
    id,
    genres,
    rating,
    runtime,
    overview = '',
    image,
  } = movie;

  return (
    <div className={styles.container}>
      { isLoading ? <Spinner /> : (
        <section className={styles.hero} style={{ backgroundImage: `url(${image})` }}>
          <div className={styles.movieInfo}>
            <h2 className={styles.title}>{ title }</h2>
            <div className={styles.wrapper}>
              <Genres genres={genres} />
              {runtime !== '0h 0m' && <span className={styles.duration}>{runtime}</span>}
            </div>
            { rating ? (
              <div className={styles.rating}>
                <StarRating
                  starsSelected={rating}
                />
                <NumRating rating={rating} />
              </div>
            ) : null
            }
          </div>
          <div className={styles.movieDescription}>
            <p className={cx({ description: showInfo, hiddenDesc: !showInfo })}>
              {`${overview.slice(0, 221)} ...`}
            </p>
            <div className={styles.buttons}>
              <Button
                type="primary"
                handleClick={
                  () => {
                    fetchVideoKey(id);
                    toggleModal(true);
                  }}
              >
                Watch Now
              </Button>
              <Button type="ghost" handleClick={() => setShowInfo(!showInfo)}>View Info</Button>
            </div>
          </div>
        </section>
      )}
    </div>
  );
};

/* istanbul ignore next */
function areEqual(prevProps, nextProps) {
  return prevProps.movie.image === nextProps.movie.image;
}

Hero.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string,
    genres: PropTypes.array,
    rating: PropTypes.number,
    runtime: PropTypes.string,
    overview: PropTypes.string,
    image: PropTypes.string,
  }).isRequired,
  fetchVideoKey: PropTypes.func.isRequired,
  toggleModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default memo(Hero, areEqual);
