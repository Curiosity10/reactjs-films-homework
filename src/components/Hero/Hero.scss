@import "../../styles/variables.scss";

.hero {
  box-sizing: border-box;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  padding-bottom: 20px;
  height: 80vh;
  position: relative;
  background: no-repeat center;
  background-size: cover;

  &::after {
    content: "";
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: linear-gradient(0deg, black, transparent 40%);
  }

  &_title {
    font-size: 2rem;
    font-family: "Helvetica Neue Bold", Arial, Helvetica, sans-serif;
    margin-bottom: 2%;
  }

  &_duration {
    border-left: 2px solid white;
    padding-left: 3%;
    font-size: 0.8rem;
    font-family: "Helvetica Neue Medium", Arial, Helvetica, sans-serif;
  }

  &_wrapper {
    margin-bottom: 20px;
  }

  &_movieInfo {
    width: 300px;
    color: white;
    margin-left: 5%;
    margin-bottom: 5px;
    z-index: 1;
  }

  &_rating {
    font-size: 0.8rem;
    display: flex;
    align-items: flex-end;

    span {
      margin-left: 3%;
    }
  }

  &_movieDescription {
    position: relative;
    z-index: 1;
    margin-right: 1%;
  }

  &_description {
    opacity: 1;
    position: relative;
    width: 250px;
    height: 95px;
    border-radius: 3px;
    background-color: rgba(0, 0, 0, 0.6);
    color: white;
    padding: 20px;
    margin-bottom: 35px;
    font-size: 0.9rem;

    &::after {
      content: '';
      border-left: 15px solid transparent;
      border-right: 15px solid transparent;
      border-top: 20px solid black;
      opacity: 0.6;
      position: absolute;
      right: 35%;
      bottom: -20px;
    }
  }

  &_hiddenDesc {
    opacity: 0;
    width: 250px;
    height: 95px;
    padding: 20px;
    margin-bottom: 35px;
  }

  &_buttons {
    display: flex;
    z-index: 1;

    & :first-child {
      margin-right: 20px;
    }
  }
}

@media all and (max-width: $tablet-screen-size) {
  .hero {
    flex-wrap: wrap;

    &_movieInfo {
      width: 150px;
      display: flex;
      flex-direction: column;
      align-items: center;

      & span:not(:last-child) {
        margin-bottom: 5px;
      }
    }

    &_duration {
      border: none;
    }

    &_wrapper {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
    }

    &_description {
      &::after {
        display: none;
      }
    }

    &_buttons {
      justify-content: center;
    }
  }
}

@media all and (max-width: $mobile-screen-size) {
  .hero {
    justify-content: center;
    align-items: center;

    &_movieInfo {
      margin-left: 0;
      margin-top: 17vh;
      width: 90vw;
      display: flex;
      flex-direction: column;
      align-items: center;
      border-radius: 3px;
      background-color: rgba(0, 0, 0, 0.6);
      padding: 5px 0;

      & span:not(:last-child) {
        margin-bottom: 5px;
      }
    }

    &_description {
      box-sizing: border-box;
      width: 90vw;
      height: auto;
    }

    &_hiddenDesc {
      box-sizing: border-box;
      width: 90vw;
      height: auto;
    }

    &_wrapper {
      display: block;
      width: 70vw;
      margin-bottom: 10px;
    }
  }
}
