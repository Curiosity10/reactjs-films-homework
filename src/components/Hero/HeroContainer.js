import { connect } from 'react-redux';
import Hero from './Hero';
import { movieDetailsSelector, isLoadingSelector } from '../../modules/app/appSelectors';
import { fetchVideoKey } from '../../modules/app/appActions';

const mapStateToProps = state => ({
  movie: movieDetailsSelector(state),
  isLoading: isLoadingSelector(state),
});

export default connect(mapStateToProps, { fetchVideoKey })(Hero);
