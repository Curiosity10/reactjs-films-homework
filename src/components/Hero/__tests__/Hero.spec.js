import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Hero from '../Hero';

describe('Hero component renders correctly', () => {
  it('Hero renders correctly', () => {
    const renderer = new ShallowRenderer();
    const result = renderer.render(
      <Hero
        movie={{}}
        fetchVideoKey={() => {}}
        toggleModal={() => {}}
        isLoading={false}
      />,
    );
    const buttons = result.props.children.props.children[1].props.children[1];

    const mockWatchNow = jest.fn(buttons.props.children[0].props.handleClick);
    const mockShowInfo = jest.fn(buttons.props.children[1].props.handleClick);

    mockWatchNow();
    mockShowInfo();

    expect(result).toMatchSnapshot();
    expect(mockWatchNow).toHaveBeenCalled();
    expect(mockShowInfo).toHaveBeenCalled();
  });

  it('Hero renders correctly with rating', () => {
    const renderer = new ShallowRenderer();
    const result = renderer.render(
      <Hero
        movie={{ title: 'Test', id: '123', rating: 4 }}
        fetchVideoKey={() => {}}
        toggleModal={() => {}}
        isLoading={false}
      />,
    );

    expect(result).toMatchSnapshot();
  });

  it('Hero renders correctly with loading', () => {
    const renderer = new ShallowRenderer();
    const result = renderer.render(
      <Hero
        movie={{}}
        fetchVideoKey={() => {}}
        toggleModal={() => {}}
        isLoading={false}
      />,
    );

    expect(result).toMatchSnapshot();
  });
});
