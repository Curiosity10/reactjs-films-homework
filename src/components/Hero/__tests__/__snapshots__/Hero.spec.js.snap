// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero component renders correctly Hero renders correctly 1`] = `
<div
  className="container"
>
  <section
    className="hero"
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  >
    <div
      className="movieInfo"
    >
      <h2
        className="title"
      />
      <div
        className="wrapper"
      >
        <Genres
          genres={Array []}
        />
        <span
          className="duration"
        />
      </div>
    </div>
    <div
      className="movieDescription"
    >
      <p
        className="description"
      >
         ...
      </p>
      <div
        className="buttons"
      >
        <Button
          handleClick={[Function]}
          type="primary"
        >
          Watch Now
        </Button>
        <Button
          handleClick={[Function]}
          type="ghost"
        >
          View Info
        </Button>
      </div>
    </div>
  </section>
</div>
`;

exports[`Hero component renders correctly Hero renders correctly with loading 1`] = `
<div
  className="container"
>
  <section
    className="hero"
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  >
    <div
      className="movieInfo"
    >
      <h2
        className="title"
      />
      <div
        className="wrapper"
      >
        <Genres
          genres={Array []}
        />
        <span
          className="duration"
        />
      </div>
    </div>
    <div
      className="movieDescription"
    >
      <p
        className="description"
      >
         ...
      </p>
      <div
        className="buttons"
      >
        <Button
          handleClick={[Function]}
          type="primary"
        >
          Watch Now
        </Button>
        <Button
          handleClick={[Function]}
          type="ghost"
        >
          View Info
        </Button>
      </div>
    </div>
  </section>
</div>
`;

exports[`Hero component renders correctly Hero renders correctly with rating 1`] = `
<div
  className="container"
>
  <section
    className="hero"
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  >
    <div
      className="movieInfo"
    >
      <h2
        className="title"
      >
        Test
      </h2>
      <div
        className="wrapper"
      >
        <Genres
          genres={Array []}
        />
        <span
          className="duration"
        />
      </div>
      <div
        className="rating"
      >
        <StarRating
          starsSelected={4}
        />
        <NumRating
          rating={4}
        />
      </div>
    </div>
    <div
      className="movieDescription"
    >
      <p
        className="description"
      >
         ...
      </p>
      <div
        className="buttons"
      >
        <Button
          handleClick={[Function]}
          type="primary"
        >
          Watch Now
        </Button>
        <Button
          handleClick={[Function]}
          type="ghost"
        >
          View Info
        </Button>
      </div>
    </div>
  </section>
</div>
`;
